Grasshopper.fs

```diff
@@ -0,0 +1,146 @@
+    let infexprV (op : string) (pxs : 'x -> Doc) (xs : seq<'x>) : Doc = 
+        let mapped = Seq.map pxs xs 
+        let resseq = Seq.map (fun x -> vsep [(String op); Indent x]) (Seq.tail mapped) 
+        hsep [Seq.head mapped; (hsep resseq)] 
+        |> parened 
+
+    // Print infix operator on one line
+    let infexpr (op : string) (pxs : 'x -> Doc) (xs : seq<'x>) : Doc = 
+        let mapped = Seq.map pxs xs 
+        let resseq = Seq.map (fun x -> hsep [(String op); x]) (Seq.tail mapped) 
+        hsep [Seq.head mapped; (hsep resseq)] 
+        |> parened 
+
+    /// Prints a marked Var 
+    let printMarkedVarGrass (v : MarkedVar) : Doc =
+        match v with 
+        | Before s -> String "before_" <-> String s 
+        | After s -> String "after_" <-> String s
+        | Intermediate (i, s) -> String "inter_" <-> String (sprintf "%A_" i) <-> String s 
+        | Goal (i, s) -> String "goal_" <-> String (sprintf "%A_" i) <-> String s 
+
+    /// Prints a typed Var 
+    let printTypedVarGrass (v : TypedVar) = 
+        match v with 
+        | Int name -> String name  
+        | Bool name -> String name  
+        | _ -> failwith "Unimplemented for Grasshopper backend." 
+
+    /// Pretty-prints an arithmetic expression.
+    let rec printIntExprG (pVar : 'Var -> Doc) : IntExpr<'Var> -> Doc =
+        function
+        | IVar c -> pVar c
+        | _ -> failwith "Unimplemented for Grasshopper backend." 
+
+    /// Pretty-prints a Boolean expression.
+    and printBoolExprG (pVar : 'Var -> Doc) : BoolExpr<'Var> -> Doc =
+        function
+        | BVar c -> pVar c
+        | BAnd xs -> infexprV "&&" (printBoolExprG pVar) xs
+        | BOr xs -> infexprV "||" (printBoolExprG pVar) xs
+        | BImplies (x, y) -> infexprV "==>" (printBoolExprG pVar)  [x; y]
+        | BNot (BEq (x, y)) -> infexpr "!=" (printExprG pVar) [x; y]
+        | BEq (x, y) -> infexpr "==" (printExprG pVar) [x; y]
+        | _ -> failwith "Unimplemented for Grasshopper backend." 
+
+    /// Pretty-prints an expression.
+    and printExprG (pVar : 'Var -> Doc) : Expr<'Var> -> Doc =
+        function
+        | Int i -> printIntExprG pVar i
+        | Bool b -> printBoolExprG pVar b
+        | _ -> failwith "Unimplemented for Grasshopper backend." 
+
+    /// Pretty-prints a symbolic sentence 
+    let rec printSymGrass (pReg : 'Reg -> Doc) (sym : Sym<'Reg>) : Doc =
+        match sym with
+        | Reg r -> pReg r
+        | Sym { Sentence = ws; Args = xs } ->
+            let pArg = printExprG (printSym pReg)
+            parened
+                (printInterpolatedSymbolicSentence pArg ws xs)
+
+    /// Get the set of accessed variables. 
+    let findVarsGrass (zterm : Backends.Z3.Types.ZTerm) : seq<MarkedVar> = 
+        // TODO @(septract) Should this conjoin the command as well? 
+        BAnd [zterm.SymBool.WPre; zterm.SymBool.Goal] 
+        |> findMarkedVars (tliftToBoolSrc (tliftToExprDest collectSymMarkedVars)) 
+        |> lift (Set.map valueOf) 
```

Comment:
Could collapse these two `lift`s into `lift (Set.map valueOf >> Set.toSeq)` but that might be overly point free